thoth/storages/graph/postgres.py

```diff
@@ -4535,37 +4566,116 @@ def sync_adviser_result(self, document: dict) -> None:
+                "adviser_name": document["metadata"]["analyzer"],
+                "adviser_version": document["metadata"]["analyzer_version"],
+                "count": parameters["count"],
+                "datetime": document["metadata"]["datetime"],
+                "debug": cli_arguments.get("verbose", False),
+                "duration": document["metadata"].get("duration"),
+                "limit": parameters["limit"],
+                "limit_latest_versions": parameters.get("limit_latest_versions"),
+                "origin": origin,
+                "is_s2i": is_s2i,
+                "recommendation_type": parameters["recommendation_type"].upper(),
+                "requirements_format": parameters["requirements_format"].upper(),
+                "external_hardware_information_id": external_hardware_info.id,
+                "external_build_software_environment_id": None,
+                "external_run_software_environment_id": external_run_software_environment.id,
+                "user_software_stack_id": software_stack.id,
+            }
 
             # Output stacks - advised stacks
             if not document["result"].get("report", {}):
                 _LOGGER.warning(
                     "No report found in %r", adviser_document_id
                 )
+                adviser_run, _ = AdviserRun.get_or_create(
+                    session,
+                    **attributes,
+                    need_re_run=need_re_run
+                )
                 return
 
+            unresolved_packages = document["result"].get("report", {}).get("_ERROR_DETAILS", {}).get("unresolved", [])
+
+            if unresolved_packages:
+                need_re_run = True
+
+            if re_run_adviser_id and unresolved_packages:
+                # If adviser was re run and there are still unsolved packages.
+                adviser_run, _ = AdviserRun.get_or_create(
+                    session,
+                    **attributes,
+                    need_re_run=need_re_run,
+                    re_run_adviser_id=re_run_adviser_id
+                )
+
+            elif re_run_adviser_id and not unresolved_packages:
+                # If adviser was re run and there are no more unsolved packages.
+
+                # Modify initial adviser flag in order to avoid re run.
+                first_adviser_run = (
+                    session.query(AdviserRun)
+                    .filter(AdviserRun.adviser_document_id == re_run_adviser_id)
+                    .first()
+                )
+                # INSERT…ON CONFLICT (Upsert)
+                # https://docs.sqlalchemy.org/en/13/dialects/postgresql.html?highlight=conflict#insert-on-conflict-upsert
+                # https://docs.sqlalchemy.org/en/13/errors.html#sql-expression-language compile required
+                if first_adviser_run and first_adviser_run.need_re_run:
+                    insert_stmt = insert(AdviserRun).values(
+                        id=first_adviser_run.id,
```

Comment:
Something like this to avoid enumerating all the properties (when they change):

```python
insert_stmt = insert(AdviserRun).values(**first_adviser_run.to_dict(without_id=False))
```

Reply:
Much better :) Thanks!